src/classes/CW_AudienceGroupController.cls

```diff
@@ -0,0 +1,543 @@
+* 0.1 01/04/2014 Cristina P.	Intial
+* 0.2 08/04/2014 Cristina P.	B-109513 - download the audience group template
+* 0.3 24/04/2014 Alex Panka   D-106803, B-110234
+*
+*/
+
+public with sharing class CW_AudienceGroupController {
+	
+	public List<CW_Audience_Group__c> 	audienceGroups {get; set;}
+	public List<audienceGroupWrapper> 	campGroups {get; set;}
+	public CW_Campaign__c 				camp {get; set;}
+	public set<Id> 						controlGroupsIds;
+	public Id							campId{get; set;}
+	public String 						contextItem {get;set;}
+	public Decimal						totalSize {get;set;}
+	public Decimal 						pageSize;
+	public Decimal 						pages;
+	public String 						deleteMessage {get; set;}
+	public String						campName {get; set;}
+	/* view mode variables */
+	public Boolean						inputMode {get;set;}
+	public Boolean						outputMode {get;set;}
+	
+	/* buttons variables */
+	public Boolean						editDisabled {get;set;}
+	public Boolean						deleteDisabled {get;set;}
+	public Boolean						saveDisabled {get;set;}
+	public Boolean						cancelDisabled {get;set;}
+	public Boolean						selectAll {get;set;}
+	public Boolean						controlGroupVisible {get;set;}
+	public Boolean						hasFirstPage {get;set;}
+	public Boolean						hasLastPage {get;set;}
+	public Boolean						editControlGroup {get; set;}
+	public Boolean						addGroup;
+	
+	public CW_CampaignListSetting__c audienceGroupPageSizeRecord;
+	public String 						taskType {get; set;}
+	
+	/* set controller variables */
+	public ApexPages.StandardSetController setGroups;
+	public Set<Id> selectedGroupIds;
+	
+	public CW_AudienceGroupController()
+	{
+		outputMode=true;
+		inputMode=false;
+		
+		//init variable
+		audienceGroupPageSizeRecord = CW_CampaignListSetting__c.getValues(ConstantUtil.AUDIENCE_GROUP_PAGE_SIZE);
+		this.selectedGroupIds= new Set<Id>();
+		campId = (Id) Apexpages.currentPage().getParameters().get('id');
+		if (campId!=null)
+		{
+			campName = getCampaignName(campId);
+			deleteMessage ='';
+			init();
+		}
+	}
+	
+	public void init()
+	{
+		//set the data display mode
+		outputMode=true;
+		inputMode=false;
+		editDisabled=false;
+		deleteDisabled = true;
+		saveDisabled = true;
+        cancelDisabled = true;
+        selectAll = false;
+        addGroup = false;
+        
+        //gather data set
+		this.setGroups= new ApexPages.StandardSetController(getAudienceSetGroups(campId));
+		this.setGroups.setpageNumber(1);
+		if (audienceGroupPageSizeRecord!=null && audienceGroupPageSizeRecord.Integer_Value__c!=null){
+			this.setGroups.setPageSize(Integer.valueOf(audienceGroupPageSizeRecord.Integer_Value__c));
+		}	else {
+			this.setGroups.setPageSize(100);
+      }
+		totalSize = this.setGroups.getResultSize();
+		
+		campGroups = getAudienceGroup();
+		getHasFirstAndLast();
+		getControlGroupIds(); 
+	}
+	
+    //list of audience group records linked to the campaign
+	public List<CW_Audience_Group__c> getAudienceSetGroups(Id campId)
+	{
+		audienceGroups = [SELECT Name, Is_Control_Group__c, 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.